level1a/handlers/level1a.py

```diff
@@ -306,6 +335,34 @@ def interpolate(
     }, index=target)
 
 
+def find_match(
+    target_date: Timestamp,
+    column: str,
+    dataframe: DataFrame,
+) -> np.ndarray:
+    matches = dataframe[
+        (dataframe["schedule_start_date"] <= target_date.asm8)
+        & (dataframe["schedule_end_date"] >= target_date.asm8)
+    ].reset_index()
+    if len(matches) > 1:
+        msg = f"Overlapping schedules for target date {target_date}"
+        raise OverlappingSchedules(msg)
+    elif len(matches) == 0:
+        msg = f"Missing schedule for target date {target_date}"
+        raise MissingSchedule(msg)
+    return matches[column][0]
+
+
+def match_with_schedule(
+    dataframe: DataFrame,
+    target: DatetimeIndex,
+) -> DataFrame:
+    return DataFrame({
+        column: [find_match(ind, column, dataframe) for ind in target]
+        for column in dataframe
+    }, index=target)
```

Comment:
här skulle jag valt apply() 💡  och ett Intervall index på schedule tabellen.

```python
from pandas import DataFrame, Series
from pandas.arrays import IntervalArray
from numpy import linspace

df1 = DataFrame(
    [1, 3, 7],
    index=IntervalArray.from_tuples([(0, 1.5), (1, 2), (2.5, 4)]),
    columns=["class"],
)
df2 = DataFrame(Series(linspace(-1, 5, 20)), columns=["data"])

def find_class(x):
    try:
        value = df1.iloc[df1.index.get_loc(x)]['class']
    except KeyError:
        return None
    return value

df2["class"] = df2["data"].apply(find_class)
```

Reply:
Såhär verkar man kunna göra:

```python
from pandas import DataFrame, Series
from pandas.arrays import IntervalArray
from numpy import linspace

df1 = DataFrame(
    [1, 3, 7],
    index=IntervalArray.from_tuples([(0, 1.5), (1, 2), (2.5, 4)]),
    columns=["class"],
)
df2 = DataFrame(Series(linspace(-1, 5, 20)), columns=["data"])

def find_class(x):
    try:
        value = df1.iloc[df1.index.get_loc(x)]['class']
    except KeyError:
        raise ValueError("No match!")
    if type(value) == Series:
        raise ValueError("Overlapping indexes!")
    return value

df2["class"] = df2["data"].apply(find_class)
```

Reply:
This looks very clever, but it doesn't quite fit how indexes are used in the rest of the code. I'll put it on the back burner for now, rather than refactor everything. May have an opportunity to come back to it soonish though.